### ISSUE TITLE
sign = no, but insists on keyring

### ISSUE THREAD

**Issue by ringods** [raphaelcohn/swaddle#25]
For the moment, I'm trying to build a single deb package with some files in it. I do not want to sign my package for the moment, so I set:

```
configure swaddle sign no
```

Swaddle still insists on the ``keyring`` being set:

```
$ ./swaddle --swaddling-path ./swaddling --output-path . -- rq-web-api-elx
swaddle: WARN: Signing disabled by default as no keyring at GNUPGHOME
swaddle: WARN: There is no Git Hub token file at '/Users/ringods/.swaddle/github-personal-access-token'
swaddle: FAIL: The configuration setting 'swaddle keyring' can not be unset
```

Can I circumvent this?

**Comment by raphaelcohn** [comment]
It's not possible in the current code base, and it's not something I'm that keen on turning off, although I do know working with GPG keys are a pain in the neck. The intent here was to try to ensure opinionated good practice post-Snowden. The validation checks for the keyring occur *before* any for signing; signing checks are both global and per packaging kind, so it's not possible to turn off the keyring check without iterating over all config, twice. It might be possible to turn this into a warning, with the result that packaging could crash mid-way leaving any GitHub related stuff in a partly uploaded mess.

Let me take a look and see if there's a halfway house that could work.

**Comment by ringods** [comment]
I just think the order of the checks should be different. Only check on everything gpg when `sign yes`.

I do agree with you on the need for properly signed packages, but at the moment it's too hard as a new user to create a package. You have to have *everything* correctly in place from the start. I should be able to increment on my package building:

* first my content
* then the base package metadata
* then more advanced metadata
* signing
* ... (did I forget something?)

`fpm` hardly does more than just putting content in a package, full Debian packaging according the policy is hard. Please let `swaddle` be the middle ground which helps me instead of blocks me.

**Comment by raphaelcohn** [comment]
I've made a change that keeps the warning but suppresses the FAIL, by creating a temporary, empty keyring. This is because a number of places expect the keyring path to be extant, and we'll get fail-fast errors for gpg if the configuration turns on signing for a packaging type without a real keyring being present. In commit 0bc7500b2c9308cf43f28b283958211f15f38d64 . This should let you proceed with no signing (so steps 1 -3), but be a reminder that that you're not yet 'secure' (ie warnings => things to fix). Likewise this should also work with RPM (where signing is really painful, because of the dumb design of rpmbuild. It's also gpg2 broken).

You're appealing to my instincts - swaddle was a reaction to fpm (and CPack) - and the need to have Ruby on my build systems (I want to be able to bootstrap from a more minimal environment - so I don't want perl, python, ruby, node or anything else that needs a huge language package manager and compiler just to work and changes so frequently it breaks - npm I'm looing at you). It is intended to produce 100% compliant Debian packages, which are of a really high quality, can exploit all the wonderful features of apt (proper upgrading, replacement, weird dependency conflicts, etc) and keep all that stuff as plain old files in source control, broken down into small pieces so merge conflicts are less likely in a big team.

Aside from signing, how hard is it do step 1 (content only) right now?

**Comment by raphaelcohn** [comment]
Tr y[this release](https://github.com/raphaelcohn/swaddle/releases/tag/release_2015.0424.1107-2)

**Comment by ringods** [comment]
@raphaelcohn after activating `multiverse` in my `trusty` docker container for the `rar` package, the installation of `swaddle` was a breeze! Thanks a lot for your time this morning!

**Comment by raphaelcohn** [comment]
Would it be worth making rar an optional dependency then? Technically, it's needed for packaging, but very few are going to use it.

**Comment by ringods** [comment]
That would be handy. Then I don't have to find a Chef way to activate the multiverse component in my build machine.

**Comment by raphaelcohn** [comment]
Let me have a think about it. The downside is that it means a fresh install of swaddle is technically broken. The other alternative would be to add `rar` to the swaddle repo. Or to make the dependencies `recommended`. Or just remove `rar` support (but that seems a bit harsh). What's best?

PS _I'm no fan of Chef, Puppet, SaltStack, Ansible or any of those other 'helpful' things, but most of the rest of the world is_. (With swaddle, you could just add in all your third-party debs from Ubuntu, and keep server config debs in source control, swaddle the lot, and you never need any of those tools ever again).

**Comment by ringods** [comment]
Well, leave it for now. The activation of `multiverse` is automated. :beers:

**Comment by ringods** [comment]
@raphaelcohn aside from signing, the next error is GitHub Release related:

```
swaddle --swaddling-path /pipeline/source/swaddling --output-path /pipeline/output -- rq-web-api-elx
swaddle: WARN: There is no Git Hub token file at '~/.swaddle/github-personal-access-token'
swaddle: WARN: Signing is disabled. Unsigned apt repositories are a very bad idea.
/usr/bin/swaddle: 26562: /usr/bin/swaddle: _swaddle_publishers_githubreleases_releaseMarkdownBody: parameter not set
```

Are you with me to do some further bug hunting so that any newcomers after me have a 5-minute first `swaddle` experience instead of a few hours? More :beers: when you ever come to Belgium. :smiley:

**Comment by raphaelcohn** [comment]
Yes would like to. I'm hors de combat now until Monday though...

Thanks for caring!

PS this is a very odd error, as it implies
`_swaddle_publishers_githubreleases_releaseMarkdownBody` is not set - this
should never be the case - and this code is executing correctly here. Hmmm.
What's the default shell on your server - dash or bash?

**Comment by ringods** [comment]
@raphaelcohn I wouldn't know wich shell is used. I use [Wercker](http://wercker.com) as my CI with my custom Docker build container. While I have set the CMD to be a login shell, I already found that this shell is not used (/etc/profile.d/ was not sourced). So I'm totally in the dark on which shell is eventually used.

@hatchan do you know technically which shell is used when you run a Docker container on Wercker to build or deploy?

**Comment by raphaelcohn** [comment]
Its probably dash. Whats /bin/sh linked to? We dont use the login shell.

**Comment by ringods** [comment]
My Docker image starts from ubuntu 14.04 which has `/bin/sh` linked to `/bin/dash`.

**Comment by stormmq** [comment]
Ta

Raphael Cohn
Chief Architect, stormmq
Co-Chair, OASIS MQTT Standard
Secretary, OASIS AMQP Standard
raphael.cohn@stormmq.com
+44 7590 675 756

UK Office:
Hamblethorpe Farm, Crag Lane, Bradley BD20 9DB, North Yorkshire, United
Kingdom
Telephone: +44 845 3712 567

Registered office:
16 Anchor Street, Chelmsford, Essex, CM2 0JY, United Kingdom
StormMQ Limited is Registered in England and Wales under Company Number
07175657
StormMQ.com

**Comment by raphaelcohn** [comment]
OK, a tiny bit of progress. The error message is that that dash, not bash would produce, and `/bin/sh` is pointing to`dash`, so that at least all tallies. Now, the odd thing is the function `swaddle_publishers_githubreleases_augmentReleaseMarkdownBody` is being called before `_swaddle_publishers_githubreleases_releaseMarkdownBody` has been set. I really need a stack trace. Could you run the swaddle in [this gist](https://gist.github.com/raphaelcohn/1c9f27be5cdf6a795f71), and put the output of stderr (`2>/path/to/some/temp/file/to/send/me`) as a gist?

(I've tried in the past to have built in debugging switches with shell scripts, but it can make them very brittle and occasionally insecure. I might take another look).

**Comment by ringods** [comment]
@raphaelcohn first of all, the shell is not `dash`, but `bash` in the Docker container:

```
CONTAINER ID        IMAGE                              COMMAND                CREATED             STATUS              PORTS               NAMES
320f67a54e18        releasequeue/ci-buildtools:0.3.0   "/bin/bash"            2 minutes ago       Up About a minute                       wercker-pipeline-924870c4-ad3d-448e-bbdc-34cd742fc5eb
```

I haven't found a way to get artifacts out of that build container and the container is removed when the build finishes, so I can't inspect it anymore afterwards. I redirected all output to a log file on the host terminal instead of the docker container: [gist](https://gist.github.com/ringods/9d2d6beda377ec786815)

**Comment by raphaelcohn** [comment]
Thank you. That gives me something to get my teeth into. I'll let you know how it goes.

**Comment by raphaelcohn** [comment]
Right, I've pushed a commit (37db6d89881f06b5019c9bc18a5f5fb3ebe9ecd8) that I believe should fix the issue. In essence, each of the methods that creates a 'repository' (be it `apt`, `yum` or just plain `files`) augments the markdown text for the release description that's used on the GitHub releases page. Because you've not set the configuration for that page, the initial setting of the markdown text wasn't occurring. Now, instead, you'll get NOTICEs (not warnings) at `-v 1`.

Before I do a release, I'm going to:-
* see if I can add a `--debug` switch
* see if I can add a `--release off` switch

**Comment by raphaelcohn** [comment]
BTW, the shell the script is executing under is `dash`. This is a cute feature in `shellfire` - it re-executes itself under a known good system shell (ie ash, dash, mksh) wherever possible. Doing this avoids any dependency on the user's login shell in `/etc/passwd` (and has nothing to do with `/etc/shells`, either), means it can run under a daemon with no shell and no home, and avoids any side effects of being run via the C function [system()](https://www.securecoding.cert.org/confluence/pages/viewpage.action?pageId=2130132&src=search) (which is a security hole waiting to happen). On Debian-derived systems, this means we use the excellent `dash` and we also work superbly on BusyBox's `sh` (which is another `ash` derivative).

**Comment by raphaelcohn** [comment]
There's a new [release](https://github.com/raphaelcohn/swaddle/releases/tag/release_2015.0428.0842-1). I've not added a `--debug` switch, as it has some complexities I'd rather avoid until I've thought things through. I have added a `--release-github` switch, and defaulted this to `off` (a change of behaviour). This also has the benefit of removing the `WARN` about a missing GitHub API token file. I've upgraded the `NOTICE` in commit 37db6d8 to a `WARN`, as these now only trigger if you're explicitly trying to publish.
